### PR TITLE
Switch to golangci-lint.

### PR DESCRIPTION
**Pull request by jpeach**
Rather than running gofmt and golint separately, switch to golangci-lint,
which runs these as well as many other linters. golangci-lint gives a
lot more control over false positive.

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        - errcheck
+        - dupl
+        - gosec
+
+    # Exclude known linters from partially hard-vendored code,
+    # which is impossible to exclude via "nolint" comments.
+    - path: internal/hmac/
+      text: "weak cryptographic primitive"
+      linters:
+        - gosec
```
Comment (hbagdi) [inline .golangci.yml:54]: Can we remove this exclude rule as it doesn't apply to this project?

File: hack/verify-golint.sh [hunk 2]
```diff
@@ -18,30 +18,13 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
+readonly VERS="v1.33.0"
```
Comment (hbagdi) [inline hack/verify-golint.sh:21]: nit: rename this to VERSION (or maybe something even more descriptive

**Review by hbagdi — APPROVED** [review]

**Comment by jpeach** [comment]
I think the CI problem is the data changes in generated files. Logs aren't particularly clear.

**Comment by hbagdi** [comment]
I _think_ this failure is because of generated files under `pkg` directory being different than in the latest commit on this PR.
Welcome to 2021!

```
 Uncommitted changes in generated sources:
 M apis/v1alpha1/zz_generated.deepcopy.go
 M apis/v1alpha1/zz_generated.register.go
 M pkg/client/clientset/versioned/clientset.go
 M pkg/client/clientset/versioned/doc.go
 M pkg/client/clientset/versioned/fake/clientset_generated.go
 M pkg/client/clientset/versioned/fake/doc.go
 M pkg/client/clientset/versioned/fake/register.go
 M pkg/client/clientset/versioned/scheme/doc.go
 M pkg/client/clientset/versioned/scheme/register.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/apis_client.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/backendpolicy.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/doc.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/fake/doc.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/fake/fake_apis_client.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/fake/fake_backendpolicy.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/fake/fake_gateway.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/fake/fake_gatewayclass.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/fake/fake_httproute.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/fake/fake_tcproute.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/fake/fake_tlsroute.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/fake/fake_udproute.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/gateway.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/gatewayclass.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/generated_expansion.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/httproute.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/tcproute.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/tlsroute.go
 M pkg/client/clientset/versioned/typed/apis/v1alpha1/udproute.go
 M pkg/client/informers/externalversions/apis/interface.go
 M pkg/client/informers/externalversions/apis/v1alpha1/backendpolicy.go
 M pkg/client/informers/externalversions/apis/v1alpha1/gateway.go
 M pkg/client/informers/externalversions/apis/v1alpha1/gatewayclass.go
 M pkg/client/informers/externalversions/apis/v1alpha1/httproute.go
 M pkg/client/informers/externalversions/apis/v1alpha1/interface.go
 M pkg/client/informers/externalversions/apis/v1alpha1/tcproute.go
 M pkg/client/informers/externalversions/apis/v1alpha1/tlsroute.go
 M pkg/client/informers/externalversions/apis/v1alpha1/udproute.go
 M pkg/client/informers/externalversions/factory.go
 M pkg/client/informers/externalversions/generic.go
 M pkg/client/informers/externalversions/internalinterfaces/factory_interfaces.go
 M pkg/client/listers/apis/v1alpha1/backendpolicy.go
 M pkg/client/listers/apis/v1alpha1/expansion_generated.go
 M pkg/client/listers/apis/v1alpha1/gateway.go
 M pkg/client/listers/apis/v1alpha1/gatewayclass.go
 M pkg/client/listers/apis/v1alpha1/httproute.go
 M pkg/client/listers/apis/v1alpha1/tcproute.go
 M pkg/client/listers/apis/v1alpha1/tlsroute.go
 M pkg/client/listers/apis/v1alpha1/udproute.go 
```

**Comment by robscott** [comment]
@jpeach all the conflicts here should have been resolved with #519. Hopefully it will be relatively easy to get this in after a rebase.

File: .golangci.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+# all available settings of specific linters
+linters-settings:
+  gofmt:
+    # simplify code: gofmt with `-s` option, true by default
+    simplify: true
+  goimports:
+    local-prefixes: sigs.k8s.io/service-apis
+  golint:
+    min-confidence: 0.9
```
Comment (robscott) [inline .golangci.yml:31]: If I'm reading correctly, the [default value here is 0.8](https://github.com/golang/lint/blob/master/golint/golint.go#L23), why set a higher value here?

File: .golangci.yml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+issues:
+  exclude-rules:
+    # Exclude some linters from running on tests files.
+    - path: _test\.go
+      linters:
+        - gocyclo
+        - errcheck
+        - dupl
+        - gosec
```
Comment (robscott) [inline .golangci.yml:47]: Why exclude these on tests?

File: .golangci.yml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  - gofmt
+  - goimports
+  - golint
+  - gosec
+  - govet
+  - misspell
+  disable-all: false
+  presets:
+  - bugs
+  - unused
```
Comment (robscott) [inline .golangci.yml:21]: There's a lot going on here and it's not clear to me why/how some of these settings were chosen. Does this originate from some kind of default or recommended configuration?

**Review by robscott — COMMENTED** [review]
Thanks for the work on this! Mostly just have some questions about how this is configured and what value it provides us with.

File: .golangci.yml (hunk 3 above)
Comment (jpeach) [reply .golangci.yml:31]: the previous script used 0.9

File: .golangci.yml (hunk 4 above)
Comment (jpeach) [reply .golangci.yml:47]: we don't care as much about these issues for tests, eg. cyclomatic complexity 🤷

File: .golangci.yml (hunk 5 above)
Comment (jpeach) [reply .golangci.yml:21]: it's a reasonable (to me) subset of checkers, culled from projects in the contour ecosystem and the golangci-lint defaults.

File: .golangci.yml (hunk 1 above)
Comment (jpeach) [reply .golangci.yml:54]: fixed

File: hack/verify-golint.sh (hunk 2 above)
Comment (jpeach) [reply hack/verify-golint.sh:21]: fixed

**Comment by robscott** [comment]
Thanks!

/lgtm

**Pull request merged**